### DISCUSSION TITLE
MultiEventProjection CustomGrouper documentation and tests are misleading

### DISCUSSION THREAD

**Discussion by migajek** [JasperFx/marten#3615] [General]
Given the fact that multistream projections are async by default [(for a very good reason, as explained in the docs)](https://martendb.io/events/projections/multi-stream-projections.html#multi-stream-projections) I'd expect the documentation example of the multistream projection with a custom grouper to **behave properly when registered as async**. 

This is however not the case, and the only reason the provided sample code actually works, is because [it's tested in the `inline` mode ](https://github.com/JasperFx/marten/blob/ecdcf75e047f846a8ee5322e9fbbbea3a344ea03/src/EventSourcingTests/Projections/MultiStreamProjections/CustomGroupers/custom_grouper_with_document_session.cs#L204)

The test does the operations in steps; each step committed separately (and thus, projection is updated each time, for the given batch only).

However in the async mode it might (and most likely **will**) **fail** - because the custom grouper [looks up the `UserFeatureToggles` documents ](https://github.com/JasperFx/marten/blob/ecdcf75e047f846a8ee5322e9fbbbea3a344ea03/src/EventSourcingTests/Projections/MultiStreamProjections/CustomGroupers/custom_grouper_with_document_session.cs#L35) - if the `UserLicenseAssigned` event end up in the same batch as the `LicenseFeatureToggled` (for given user/license), the query returns no results and therefore the grouper can't figure out which `UserFeatureToggles` documents should be affected.

BTW, isn't the grouper doing exactly what [the docs warn us against?](https://martendb.io/events/projections/multi-stream-projections.html#view-projection-with-custom-grouper)

> If your grouping logic requires you to access the aggregate view itself, ViewProjection will not function correctly 

**conclusion**
If I'm not mistaken, the scenario provided in the tests is still not solvable with the custom grouper (unless we set BatchSize to 1)?

Inspired by a recent question on the discord, I've crafted a solution that - instead of looking up the documents (which is the source of the issue), looks up the events - combines the past events with current batch and does a simple aggregation. 
While it technically works, I don't think this is a good solution, it might get very tricky for more complex aggregations, and I'm not even sure whether this *feels* right.

```
var licenseIdsDistinct = licenseIds.Distinct().ToArray();
var commitedAssignments = await session.Events
    .QueryRawEventDataOnly<UserLicenseAssigned>()
    .Where(x => x.LicenseId.IsOneOf(licenseIds))
    .ToListAsync();
var uncommitedAssignments = events
    .OfType<IEvent<UserLicenseAssigned>>()
    .Select(x => x.Data)
    .Where(x => x.LicenseId.IsOneOf(licenseIdsDistinct))
    .ToArray();
var allEvents = commitedAssignments.Concat(uncommitedAssignments);
var streamIds = allEvents.GroupBy(x => x.LicenseId, x => x.UserId)
    .ToDictionary(x => x.Key, x => x.ToArray());
```

**Comment by jeremydmiller** [comment]
@migajek Sorry, I don't scan the discussions very closely -- hence the pinned discussion urging you to use Discord instead.

That's a really, really old example that hasn't been touched in a long time. By and large, I think you'll have bad performance from having to query back at the old events to do grouping like that. You might try for some kind of multi-stage thing where you do a lightweight Inline project that just writes the relationships to a document or even flat table, then use that in a custom slicer so you can batch the lookups. And maybe even cache the relationship lookup

**Comment by migajek** [comment]
@jeremydmiller thanks, I am aware of the Discord, however I thought it's good for prompt subjects. This is by no means a high priority thing, so decided to raise it here instead.

I agree on the performance penalty of querying events for aggregation, and - personally, after approaching the problem in a different ways, decided to go with the "helper" projection that is inline. 

What I believe is a missing piece of puzzle is some kind of marking the dependency between projections. Rebuilding the main projection without ensuring the helper projection is rebuilt first (in case of any changes to the logic, etc) will result in integrity issues. 
I handle it manually now, that is - the documentation, the comments in the code etc, but I wonder whether you considered any mechanisms in Marten itself (or the command line tooling) to mark the dependencies between projections

**Reply by jeremydmiller** [reply]
See this: https://github.com/JasperFx/marten/issues/3703

Long story short, this hasn't been done yet because its' just not a quick thing to do. There's a JasperFx client (or 2) that will benefit from it, and might sponsor some of that work in the near-ish future

**Comment by ghord** [comment]
Aren't more examples broken in docs? I'm trying to write correct multi stream projection while learning marten, but to me it seems every Example in Custom Aggregate grouping is subtly wrong:

For example in: *Pattern 1, resolve the aggregate id through an inline lookup projection*:

What happens if we have events in a batch like
[
   ShippingLabelCreated,
   CustomerLinkedToExternalAccount
]

Grouping logic could see customer account linked **before** shipping label created, since it is using inline projection for linked accounts processing. This would cause multi-stream projection to receive the event on not-yet-linked account which is incorrect.

*Pattern 2, keep the linked single stream ids on the projected document, then query by containment*

This example seems to be using invalid pattern that is warned against in the documentation (querying projection itself inside grouping logic). So for example if we get batch 

[
   CustomerLinkedToExternalAccount
   ShippingLabelCreated,
]

We will miss shipping label, because at the time of grouping the document will not yet be linked.

At this point I'm starting to doubt if it is possible to actually write correct multi-stream projections due to batching issues.  What worries me is an absolute complexity of writing something correctly when it is bascially a simple JOIN statement in sql in relational database model.

**Reply by jeremydmiller** [reply]
@ghord I don't know what to tell you here, I interact with people every single day who are happily using MultiStreamProjections with Marten quite successfully. Do you have a specific question, or suggestion, or just griping in general? 

And maybe go look at the documentation on "Composite Projections" to get the timing right.

**Reply by ghord** [reply]
Are examples in the docs correct and resilient to the edge cases I mentioned? Or am I misunderstanding how batching works?

**Reply by jeremydmiller** [reply]
I think you're misunderstanding the batching. You have access to the entire batch at one time in the grouper. I'm not sure why you think none of this works. And the composite projection stuff came after those examples were written, so that might be where I'd go. We have plenty of samples that clearly demonstrate how you could create denormalized views.

**Reply by ghord** [reply]
I went and created a simple repro of failures of both examples I'm mentioning. I can reproduce both tests failing due to batching issues. I simply pasted the classes from the examples on the website and added test cases for both that illustrate the issue that final projection state depends on whether the event are processed in a single batch or in multiple batches.

https://gist.github.com/ghord/8ed794e27f2757d2a569ac1154b8bea6

I admit I had not tested Composite Projections for this, I'll take a look, thank you for suggestion.

**Reply by jeremydmiller** [reply]
Okay. I'll look probably tomorrow.
